src/server/pfs/server/driver.go

```diff
@@ -1729,14 +1755,86 @@ func (d *driver) squashCommitSet(txnCtx *txncontext.TransactionContext, commitse
 			return &pfsserver.ErrSquashWithoutChildren{Commit: ci.Commit}
 		}
 	}
-
 	if err := d.squashCommitSetInternal(txnCtx, commitInfos); err != nil {
 		return err
 	}
+	for _, ci := range commitInfos {
+		txnCtx.StopJob(ci.Commit)
+	}
+	return nil
+}
 
-	// notify PPS that this commitset has been squashed so it can clean up any
-	// jobs associated with it at the end of the transaction
-	txnCtx.StopJobs(commitset)
+// getCommitSubvenance gets the subvenance of the provided commit.
+// The provided commit is included in the returned list.
+// There are no guarantees about the ordering of the returned list.
+func (d *driver) getCommitSubvenance(txnCtx *txncontext.TransactionContext, commit *pfs.Commit) ([]*pfs.CommitInfo, error) {
+	commitInfos, err := d.inspectCommitSetImmediate(txnCtx, &pfs.CommitSet{ID: commit.ID})
+	if err != nil {
+		return nil, err
+	}
+	// Create an index from branch -> commit for direct provenance lookups.
+	branchIndex := make(map[string]*pfs.CommitInfo)
+	for _, ci := range commitInfos {
+		branchIndex[pfsdb.BranchKey(ci.Commit.Branch)] = ci
+	}
+	visited := make(map[string]*pfs.CommitInfo)
+	// Traverse will fill the visited map with commits that are traversed.
+	// Non-nil commits are in the subvenance.
+	var traverse func(*pfs.CommitInfo)
+	traverse = func(commitInfo *pfs.CommitInfo) {
+		if _, ok := visited[pfsdb.CommitKey(commitInfo.Commit)]; ok {
+			return
+		}
+		if proto.Equal(commitInfo.Commit, commit) {
+			visited[pfsdb.CommitKey(commitInfo.Commit)] = commitInfo
+			return
+		}
+		visited[pfsdb.CommitKey(commitInfo.Commit)] = nil
+		for _, provBranch := range commitInfo.DirectProvenance {
+			ci := branchIndex[pfsdb.BranchKey(provBranch)]
+			traverse(ci)
+			ci = visited[pfsdb.CommitKey(ci.Commit)]
+			if ci != nil {
+				visited[pfsdb.CommitKey(commitInfo.Commit)] = commitInfo
+				return
+			}
+		}
+	}
+	for _, ci := range commitInfos {
+		traverse(ci)
+	}
+	var result []*pfs.CommitInfo
+	for _, ci := range visited {
+		if ci != nil {
+			result = append(result, ci)
+		}
+	}
+	return result, nil
+}
+
+func (d *driver) dropCommit(txnCtx *txncontext.TransactionContext, commit *pfs.Commit) error {
+	// Get the subvenant commits of the provided commit.
+	commitInfos, err := d.getCommitSubvenance(txnCtx, commit)
+	if err != nil {
+		return err
+	}
+	// NOTE: These checks should be unecessary, but they are included to ensure our assumptions are correct.
+	for _, ci := range commitInfos {
```

Comment:
with the old model, I believe there are still scenarios where some of the commits in a commit set will have children and some won't. If we hit such scenarios, it might be very tricky to untangle their commit graph, even with the DropCommit & SquashCommit methods exposed

Reply:
In the case that two commit sets have overlapping branches, we may have a scenario where a commit on the overlapping branches get children commits via ALIAS propagation and now a commit set has some commits with children and some without